modules/pam_timestamp/hmac_openssl_wrapper.c

```diff
@@ -81,22 +86,39 @@ get_crypto_algorithm(pam_handle_t *pamh, int debug){
 }
 
 static int
+PAM_NONNULL((1, 2))
 generate_key(pam_handle_t *pamh, char **key, size_t key_size)
 {
     int fd = 0;
     ssize_t bytes_read = 0;
-    char * tmp = NULL;
+    char *tmp = *key = NULL;
 
-    fd = open("/dev/urandom", O_RDONLY);
-    if (fd == -1) {
-        pam_syslog(pamh, LOG_ERR, "Cannot open /dev/urandom: %m");
+    tmp = calloc(key_size, 1);
+    if (!tmp) {
+        pam_syslog(pamh, LOG_CRIT, "Not enough memory");
         return PAM_AUTH_ERR;
     }
 
-    tmp = malloc(key_size);
-    if (!tmp) {
-        pam_syslog(pamh, LOG_CRIT, "Not enough memory");
-        close(fd);
+    //Try to get random data from OpenSSL first
+    if (RAND_priv_bytes((unsigned char *)tmp, key_size) == 1) {
+        *key = tmp;
+        return PAM_SUCCESS;
+    }
+
+#ifdef HAVE_GETRANDOM
+    // Fallback to getrandom(5) if available
+    if (getrandom(tmp, key_size, 0) == (ssize_t)key_size) {
+        *key = tmp;
+        return PAM_SUCCESS;
+    }
+#endif
+
+    // Fallback to /dev/urandom
+    fd = open("/dev/urandom", O_RDONLY);
+    if (fd == -1) {
+        pam_syslog(pamh, LOG_ERR, "Cannot open /dev/urandom: %m");
+        pam_overwrite_n(tmp, key_size);
```

Comment:
This particular zeroing is applied to an area of memory where no data has been written yet.

Reply:
This zeroing applies to an area, where previous operations (`RAND_priv_bytes`/`getrandom`) may have partially filled the block.